src/osd/scrubber/pg_scrubber.cc

```diff
@@ -2458,117 +2401,105 @@ namespace Scrub {
 {
-  auto m = new MOSDScrubReserve(spg_t(m_pg_info.pgid.pgid, peer.shard),
-				epoch,
-				MOSDScrubReserve::RELEASE,
-				m_pg->pg_whoami);
-  m_osds->send_message_osd_cluster(peer.osd, m, epoch);
+  auto m = make_message<MOSDScrubReserve>(
+      spg_t{m_pgid.pgid, peer.shard}, epoch, MOSDScrubReserve::RELEASE,
+      m_whoami);
+  m_pg->send_cluster_message(peer.osd, m, epoch, false);
 }
 
-ReplicaReservations::ReplicaReservations(
-  PG* pg,
-  pg_shard_t whoami,
-  Scrub::ScrubJobRef scrubjob,
-  const ConfigProxy& conf)
-    : m_pg{pg}
-    , m_acting_set{pg->get_actingset()}
+void ReplicaReservations::send_request_to_replica(
+    pg_shard_t peer,
+    epoch_t epoch)
+{
+  auto m = make_message<MOSDScrubReserve>(
+      spg_t{m_pgid.pgid, peer.shard}, epoch, MOSDScrubReserve::REQUEST,
+      m_whoami);
+  m_pg->send_cluster_message(peer.osd, m, epoch, false);
+  m_requests_sent++;
+  m_request_sent_at = clock::now();
+  dout(10) << fmt::format(
+		  "{}: reserving {} ({} of {})", __func__, *m_next_to_request,
+		  m_requests_sent, m_total_needeed)
+	   << dendl;
+}
+
+ReplicaReservations::ReplicaReservations(ScrubMachineListener& scrbr)
+    : m_scrubber{scrbr}
+    , m_pg{m_scrubber.get_pg()}
+    , m_cct{m_scrubber.get_pg_cct()}
+    , m_whoami{m_pg->pg_whoami}
+    , m_pgid{m_scrubber.get_spgid()}
     , m_osds{m_pg->get_pg_osd(ScrubberPasskey())}
-    , m_pending{static_cast<int>(m_acting_set.size()) - 1}
-    , m_pg_info{m_pg->get_pg_info(ScrubberPasskey())}
-    , m_scrub_job{scrubjob}
-    , m_conf{conf}
+    , m_conf{m_cct->_conf}
 {
-  epoch_t epoch = m_pg->get_osdmap_epoch();
+  const epoch_t epoch = m_pg->get_osdmap_epoch();
   m_log_msg_prefix = fmt::format(
-      "osd.{} ep: {} scrubber::ReplicaReservations pg[{}]: ", m_osds->whoami,
-      epoch, pg->pg_id);
-
-  m_timeout = conf.get_val<std::chrono::milliseconds>(
-      "osd_scrub_slow_reservation_response");
+      "osd.{} ep: {} scrubber::ReplicaReservations pg[{}]: ", m_whoami.osd,
+      epoch, m_pgid);
+
+  // sort the acting set, so that we send the requests in a consistent order
+  // (reducing the chance of having two PGs that share some of their acting-set
+  // OSDs, consistently interfering with each other's reservation process)
+  auto acting = m_pg->get_actingset();
+  m_sorted_secondaries.reserve(acting.size());
+  std::copy_if(
+      acting.cbegin(), acting.cend(), std::back_inserter(m_sorted_secondaries),
+      [whoami=m_whoami](const pg_shard_t& shard) { return shard != whoami; });
+
+  // sorted by OSD number
+  std::sort(
+      m_sorted_secondaries.begin(), m_sorted_secondaries.end(),
+      [](const pg_shard_t& a, const pg_shard_t& b) { return a.osd < b.osd; });
+  m_total_needeed = m_sorted_secondaries.size();
+  dout(10) << fmt::format("{}: acting: {}", __func__, m_sorted_secondaries)
+	   << dendl;
 
-  if (m_pending <= 0) {
+  m_next_to_request = m_sorted_secondaries.cbegin();
```

Comment:
This next bit is semantically identical to handle_reserve_grant -- can we pull it out into a helper send_next_reservation_or_complete().